core/src/main/scala/cats/data/Kleisli.scala

```diff
@@ -107,13 +106,43 @@ final case class Kleisli[F[_], -A, B](run: A => F[B]) { self =>
   def tap[AA <: A](implicit F: Functor[F]): Kleisli[F, AA, AA] =
     Kleisli(a => F.as(run(a), a))
 
-  /** Yield computed B combined with input value. */
+  /**
+   * Yield computed B combined with input value.
+   *
+   * {{{
+   * scala> import cats.data.Kleisli, cats.implicits._
+   * scala> val base = Kleisli((i: Int) => (i + 1).pure[Option])
+   * scala> base.tapWith[String, Int]((a, b) => (a, b).toString).run(42)
+   * res0: Option[String] = Some((42,43))
+   * }}}
+   */
   def tapWith[C, AA <: A](f: (AA, B) => C)(implicit F: Functor[F]): Kleisli[F, AA, C] =
     Kleisli(a => F.map(run(a))(b => f(a, b)))
 
+  /**
+   * {{{
+   * scala> import cats.data.Kleisli, cats.implicits._
+   * scala> val base = Kleisli((i: Int) => (i + 1).pure[Option])
+   * scala> base.tapWithF[String, Int]((a, b) => Option((a, b).toString)).run(42)
+   * res0: Option[String] = Some((42,43))
+   * }}}
+   */
   def tapWithF[C, AA <: A](f: (AA, B) => F[C])(implicit F: FlatMap[F]): Kleisli[F, AA, C] =
     Kleisli(a => F.flatMap(run(a))(b => f(a, b)))
 
+  /**
+   * Yield computed B combined with input value.
+   *
+   * {{{
+   * scala> import cats.data.Kleisli, cats.implicits._
+   * scala> val base = Kleisli((_: Int).toString.pure[Either[Throwable, *]])
+   * scala> base.tapWith[Int]().run(42)
+   * res0: scala.util.Either[Throwable,(Int, String)] = Right((42,42))
+   * }}}
+   */
+  def tapWith[AA <: A]()(implicit F: Functor[F]): Kleisli[F, AA, (AA, B)] =
```

Comment:
I'm not a fan of overloading functions in this way. Often the type inference will struggle more when doing so. Maybe we can find a different name? `tapProduct`? I don't know :smile:

Reply:
Honestly I'd prefer to drop these and go with the enhancement method implementation anyway, where I named the functions `tapWithIdentity`. I'm open to using `tapProduct` there though.